internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -0,0 +1,363 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	bmh_v1alpha1 "github.com/metal3-io/baremetal-operator/pkg/apis/metal3/v1alpha1"
+	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
+	"github.com/openshift/assisted-service/models"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// BMACReconciler reconciles a Agent object
+type BMACReconciler struct {
+	client.Client
+	Log    logrus.FieldLogger
+	Scheme *runtime.Scheme
+}
+
+const (
+	AGENT_BMH_LABEL                 = "adi.openshift.io/bmh"
+	BMH_AGENT_ROLE                  = "bmac.adi.openshift.io/role"
+	BMH_AGENT_HOSTNAME              = "bmac.adi.openshift.io/hostname"
+	BMH_AGENT_MACHINE_CONFIG_POOL   = "bmac.adi.openshift.io/mcp"
+	BMH_INSTALL_ENV_LABEL           = "installenvs.adi.openshift.io"
+	BMH_INSPECT_ANNOTATION          = "inspect.metal3.io"
+	BMH_HARDWARE_DETAILS_ANNOTATION = "inspect.metal3.io/hardwaredetails"
+)
+
+// +kubebuilder:rbac:groups=metal3.io,resources=baremetalhosts,verbs=get;list;watch;update;patch
+
+func (r *BMACReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	bmh := &bmh_v1alpha1.BareMetalHost{}
+
+	if err := r.Get(ctx, req.NamespacedName, bmh); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	agent := r.findAgent(ctx, bmh)
+
+	// handle multiple agents matching the
+	// same BMH's Mac Address
+
+	if agent != nil {
+		if !agent.Spec.Approved {
+			r.reconcileAgentSpec(bmh, agent)
+		}
+
+		if agent.Status.Inventory.ReportTime != nil {
+			r.reconcileAgentInventory(bmh, agent)
+		}
```

Comment:
you should probably only update the agent once. so these functions should return (bool - has it been changed, error)